tracer.go

```diff
@@ -158,25 +158,39 @@ func (t *tracer) startSpanWithOptions(
 		options.StartTime = t.timeNow()
 	}
 
+	var references []Reference
 	var parent SpanContext
 	var hasParent bool // need this because `parent` is a value, not reference
+	var followsFromContext SpanContext
 	for _, ref := range options.References {
+		ctx, ok := ref.ReferencedContext.(SpanContext)
+		if !ok {
+			t.logger.Error(fmt.Sprintf(
+				"Reference contains invalid type of SpanReference: %s",
+				reflect.ValueOf(ref.ReferencedContext)))
+			continue
+		}
 		if ref.Type == opentracing.ChildOfRef {
-			if p, ok := ref.ReferencedContext.(SpanContext); ok {
-				if p.IsValid() || p.isDebugIDContainerOnly() || len(p.baggage) != 0 {
-					parent = p
-					hasParent = true
-					break
-				}
-			} else {
-				t.logger.Error(fmt.Sprintf(
-					"ChildOf reference contains invalid type of SpanReference: %s",
-					reflect.ValueOf(ref.ReferencedContext)))
+			references = append(references, Reference{Type: opentracing.ChildOfRef, Context: ctx})
+			if (ctx.IsValid() || ctx.isDebugIDContainerOnly() || len(ctx.baggage) != 0) && !hasParent {
+				parent = ctx
+				hasParent = true
+			}
+		} else if ref.Type == opentracing.FollowsFromRef {
+			references = append(references, Reference{Type: opentracing.FollowsFromRef, Context: ctx})
+			if (ctx.IsValid() || ctx.isDebugIDContainerOnly() || len(ctx.baggage) != 0) && !followsFromContext.IsValid() {
+				// Keep track of the first valid FollowsFrom reference in case there was no
+				// parent span
+				followsFromContext = ctx
 			}
-		} else {
-			// TODO support other types of span references
 		}
 	}
+	if !hasParent && followsFromContext.IsValid() {
+		// If ChildOfRef wasn't found but a FollowFromRef exists, use the context from
+		// the FollowFromRef as the parent
+		hasParent = true
+		parent = followsFromContext
+	}
```

Comment:
I think this can be simplified. I'd rather not have a separate code branch for every ref type.